pkg/fleet-manager/application_controller.go

```diff
@@ -0,0 +1,384 @@
+/*
+Copyright Kurator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fleet
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	helmv2b1 "github.com/fluxcd/helm-controller/api/v2beta1"
+	kustomizev1 "github.com/fluxcd/kustomize-controller/api/v1"
+	sourcev1 "github.com/fluxcd/source-controller/api/v1"
+	sourcev1a2 "github.com/fluxcd/source-controller/api/v1beta2"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apiserrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/cluster-api/util/patch"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	applicationapi "kurator.dev/kurator/pkg/apis/apps/v1alpha1"
+	fleetapi "kurator.dev/kurator/pkg/apis/fleet/v1alpha1"
+)
+
+const (
+	GitRepoKind       = sourcev1.GitRepositoryKind
+	HelmRepoKind      = sourcev1a2.HelmRepositoryKind
+	OCIRepoKind       = sourcev1a2.OCIRepositoryKind
+	KustomizationKind = kustomizev1.KustomizationKind
+	HelmReleaseKind   = helmv2b1.HelmReleaseKind
+
+	ApplicationLabel     = "apps.kurator.dev/app-name"
+	AppsPolicyLabel      = "apps.kurator.dev/policy-name"
+	ApplicationKind      = "Application"
+	ApplicationFinalizer = "apps.kurator.dev"
+)
+
+// ApplicationManager reconciles an Application object
+type ApplicationManager struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (a *ApplicationManager) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		For(&applicationapi.Application{}).
+		Build(a)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Consider whether it's necessary to watch the fleet/cluster/attachedCluster.
```

Comment:
I think we donot need to watch per cluster as when cluster is created, the relevant fleet status will be updated

Reply:
Maybe we need to know which clusters are ready

Reply:
then we add this watch